File: programs/sdk/assemblyscript/assembly/event.ts
/**
 * Event bindings.
 *
 * Events are emitted under the calling program's namespace and the invoking
 * principal. Topic and payload bounds are checked before the host call, so an
 * oversized event never reaches the runtime.
 */

import { MAX_EVENT_DATA_BYTES, MAX_EVENT_TOPIC_BYTES } from "./abi";
import { pointer } from "./bytes";
import { ERR_DATA_TOO_LARGE, ERR_EMPTY_TOPIC, ERR_TOPIC_TOO_LARGE } from "./error";
import { eventEmit } from "./host";

/** Emits one event under this program's namespace. */
export function emitEvent(topic: StaticArray<u8>, data: StaticArray<u8>): i32 {
  if (topic.length == 0) return ERR_EMPTY_TOPIC;
  if (topic.length > MAX_EVENT_TOPIC_BYTES) return ERR_TOPIC_TOO_LARGE;
  if (data.length > MAX_EVENT_DATA_BYTES) return ERR_DATA_TOO_LARGE;
  return eventEmit(pointer(topic), topic.length, pointer(data), data.length);
}

File: programs/sdk/rust/src/event.rs
//! Event bindings.
//!
//! Events are emitted under the calling program's namespace and the invoking
//! principal. Topic and payload bounds are checked at construction, so an
//! oversized event never reaches the host.

use crate::abi::{MAX_EVENT_DATA_BYTES, MAX_EVENT_TOPIC_BYTES};
use crate::error::{Field, ProgramError, Reason};

#[cfg(target_arch = "wasm32")]
use crate::host;

/// The topic one emitted event is filed under.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct EventTopic<'a>(&'a [u8]);

impl<'a> EventTopic<'a> {
    /// Constructs a topic inside the version-one event bound.
    ///
    /// # Errors
    ///
    /// Refuses an empty topic and a topic past the declared bound.
    pub const fn new(bytes: &'a [u8]) -> Result<Self, ProgramError> {
        if bytes.is_empty() {
            return Err(ProgramError::value(Field::EventTopic, Reason::Empty));
        }
        if bytes.len() > MAX_EVENT_TOPIC_BYTES {
            return Err(ProgramError::value(Field::EventTopic, Reason::TooLarge));
        }
        Ok(Self(bytes))
    }

    /// Borrows the canonical topic bytes.
    #[must_use]
    pub const fn bytes(self) -> &'a [u8] {
        self.0
    }
}

/// The payload one emitted event carries.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct EventData<'a>(&'a [u8]);

impl<'a> EventData<'a> {
    /// Constructs a payload inside the version-one event bound.
    ///
    /// # Errors
    ///
    /// Refuses a payload past the declared bound.
    pub const fn new(bytes: &'a [u8]) -> Result<Self, ProgramError> {
        if bytes.len() > MAX_EVENT_DATA_BYTES {
            return Err(ProgramError::value(Field::EventData, Reason::TooLarge));
        }
        Ok(Self(bytes))
    }

    /// Borrows the canonical payload bytes.
    #[must_use]
    pub const fn bytes(self) -> &'a [u8] {
        self.0
    }
}

/// Emits one event under this program's namespace.
///
/// # Errors
///
/// Refuses missing emit authority and every bound the host enforces.
#[cfg(target_arch = "wasm32")]
pub fn emit(topic: EventTopic<'_>, data: EventData<'_>) -> Result<(), ProgramError> {
    host::event_emit(topic.bytes(), data.bytes())?;
    Ok(())
}
